### PR TITLE
Issue 1215 partitionBy

### PR DESCRIPTION
**Pull request by rnjtranjan**

### CODE REVIEW

**Comment by harendra-kumar** [comment]
Can you rebase it and use unsafeSwapIndices?

**Comment by harendra-kumar** [comment]
Can you add a benchmark for this? I want to try out a few perf experiments.

File: benchmark/streamly-benchmarks.cabal [hunk 1]
```diff
@@ -483,6 +483,17 @@ benchmark Data.Array.Foreign
   if flag(limit-build-mem)
     ghc-options: +RTS -M1000M -RTS
 
+benchmark Data.Mut.Array
```
Comment (harendra-kumar) [inline benchmark/streamly-benchmarks.cabal:486]: Benchmark modules are usually named after the library modules. This should be named `Data.Array.Foreign.Mut`.

File: benchmark/streamly-benchmarks.cabal [hunk 2]
```diff
@@ -483,6 +483,17 @@ benchmark Data.Array.Foreign
...
+  import: bench-options
+  type: exitcode-stdio-1.0
+  hs-source-dirs: Streamly/Benchmark/Data/Mut
```
Comment (harendra-kumar) [inline benchmark/streamly-benchmarks.cabal:489]: This should be in Streamly/Benchmark/Data/Array/Foreign/Mut .

File: src/Streamly/Internal/Data/Array/Stream/Mut/Foreign.hs [hunk 3]
```diff
@@ -21,6 +21,8 @@ module Streamly.Internal.Data.Array.Stream.Mut.Foreign
     , compactLE
     , compactEQ
     , compactGE
+    , Array(..)
+    , partitionBy
```
Comment (harendra-kumar) [inline src/Streamly/Internal/Data/Array/Stream/Mut/Foreign.hs:25]: Why is this added?

**Review by harendra-kumar — CHANGES_REQUESTED** [review]

**Review by rnjtranjan — COMMENTED** [review]
looks good.

**Comment by harendra-kumar** [comment]
@rnjtranjan can you push a commit fixing the above comments.

File: src/Streamly/Internal/Data/Array/Stream/Mut/Foreign.hs (hunk 3 above)
Comment (rnjtranjan) [reply src/Streamly/Internal/Data/Array/Stream/Mut/Foreign.hs:25]: used in Streamly/Benchmark/Data/Array/Foreign/Mut/Array.hs

Comment (harendra-kumar) [reply src/Streamly/Internal/Data/Array/Stream/Mut/Foreign.hs:25]: Why is a mutable array API being exported via ArrayStream module and then used from there? Why aren't we using it directly from the mutable array module?

File: benchmark/Streamly/Benchmark/Data/Mut/Array.hs [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    main
+  ) where
+
+import Control.DeepSeq (NFData(..))
+import System.Random (randomRIO)
+import Prelude hiding ()
+
+import qualified Streamly.Prelude  as Stream
+import qualified Streamly.Internal.Data.Fold as Fold
+import qualified Streamly.Internal.Data.Array.Stream.Mut.Foreign as MArray
```
Comment (harendra-kumar) [inline benchmark/Streamly/Benchmark/Data/Mut/Array.hs:32]: We should import the array APIs from `Data.Array.Foreign.Mut`.

File: benchmark/Streamly/Benchmark/Data/Mut/Array.hs [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    :: NFData b
+    => String -> (Int -> t IO a) -> (t IO a -> IO b) -> Benchmark
+benchIO name src sink =
+    bench name $ nfIO $ randomRIO (1,1) >>= sink . src
+
+{-# INLINE partitionBy #-}
+partitionBy :: Int -> SerialT IO (MArray.Array Int) -> IO ()
+partitionBy value s = Stream.fold Fold.drain $ do
+    a <- s
+    MArray.partitionBy ( < value) a
```
Comment (harendra-kumar) [inline benchmark/Streamly/Benchmark/Data/Mut/Array.hs]: This is unnecessarily complicated. We should run the benchmark on the array itself instead of running it on a stream of arrays.

File: src/Streamly/Internal/Data/Array/Stream/Mut/Foreign.hs (hunk 3 above)
Comment (harendra-kumar) [reply src/Streamly/Internal/Data/Array/Stream/Mut/Foreign.hs:29]: This is the wrong module to export these from. The right module is `Data.Array.Foreign.Mut` and these are already exported from that module. So we can just remove these from here.

File: benchmark/Streamly/Benchmark/Data/Mut/Array.hs (hunk 4 above)
Comment (rnjtranjan) [reply benchmark/Streamly/Benchmark/Data/Mut/Array.hs]: arraysOf :: (IsStream t, MonadIO m, Storable a)
    => Int -> t m a -> t m (Array a) is not defined here (Data.Array.Foreign.Mut). Do we need to redefine here?

Comment (harendra-kumar) [reply benchmark/Streamly/Benchmark/Data/Mut/Array.hs]: We should not be using arraysOf, see my other comment. And even if we need that we can import that from this module and import other stuff from mutable array module.

File: benchmark/Streamly/Benchmark/Data/Mut/Array.hs (hunk 5 above)
Comment (rnjtranjan) [reply benchmark/Streamly/Benchmark/Data/Mut/Array.hs]: agreed.

File: benchmark/Streamly/Benchmark/Data/Array/Foreign/Mut/Array.hs [hunk 6]
```diff
@@ -57,20 +56,19 @@ sourceUnfoldrM value n = Stream.unfoldrM step n
...
+    => String -> (Int -> [a]) -> ([a] -> IO b) -> Benchmark
 benchIO name src sink =
     bench name $ nfIO $ randomRIO (1,1) >>= sink . src
 
 {-# INLINE partitionBy #-}
-partitionBy :: Int -> SerialT IO (MArray.Array Int) -> IO ()
-partitionBy value s = Stream.fold Fold.drain $ do
-    a <- s
-    MArray.partitionBy ( < value) a
+partitionBy :: Int -> [MArray.Array Int] -> IO ()
```
Comment (harendra-kumar) [inline benchmark/Streamly/Benchmark/Data/Array/Foreign/Mut/Array.hs:64]: Why list of arrays? We need just a single array.

**Review by harendra-kumar — COMMENTED** [review]
You can use change this:

```
Stream.toList $ MArray.arraysOf value $ sourceUnfoldrM value 0
```

to something like

```
Array.fromStream $ sourceUnfoldrM value 0
```
